jax/lax/lax_control_flow.py

```diff
@@ -586,21 +589,23 @@ def _scan_partial_eval(trace, *tracers, **kwargs):
   num_xs = len(jaxpr.in_avals) - num_carry - num_consts
   num_ys = len(jaxpr.out_avals) - num_carry
 
-  unknowns = original_unknowns = [t.pval[0] is not None for t in tracers]
+  unknowns = [t.pval[0] is not None for t in tracers]
   const_uk, init_uk, xs_uk = split_list(unknowns, [num_consts, num_carry])
 
+  # Fixpoint computation of which carry are unknown (not a constant): either
+  # unknown from init, or the carry out is unknown. Each iteration promotes
+  # at least one carry to unknown.
   carry_uk = init_uk
-  for _ in range(1000):
+  while True:
     unknowns = const_uk + carry_uk + xs_uk
     jaxpr_1, jaxpr_2, out_uk = pe.partial_eval_jaxpr(
         jaxpr, unknowns, instantiate=carry_uk + [False] * num_ys)
```

Comment:
Wondering if instantiate=False would be sufficient here.

Reply:
Good question. I think we tried that, and it's not, to ensure monotonicity: we need to ensure things are at least as lifted as they were on the previous iteration. For example, consider the carry being a pair, the initial carry has one unknown and one known, and the body swaps the elements of the carry. That would iterate forever, with `carry_uk` switching between `[True, False]` and `[False, True]` IIUC. Do you buy that argument?

Reply:
I do not understand well what "instantiate" does. I think it is correct the way it is now, but not from a termination perspective. The termination is ensured by the fact that in each iteration we OR the before and after; we also check for equality. This ensures that in each iteration we move monotonically towards all-true. This ensures termination. The instantiate should be passed, because I presume it will affect the jaxpr_1 and jaxpr_2 which are used below.

Reply:
`instantiate` does the OR itself on [this line in `trace_to_subjaxpr`](https://github.com/google/jax/blob/3e9ce2f69f118d89fde6e3b18f9c4a8da26a2764/jax/interpreters/partial_eval.py#L357), which calls `instantiate_const_at` and ultimately `full_raise` to ensure that an output is lifted into the trace if the corresponding entry in `instantiate` is True, i.e. so that an output is either already in the trace OR its entry of `instantiate` is True. So we agree an OR must be done, I'm just pointing out that it's already done by `instantiate`.

I may have written this comment without thinking about the explicit OR also added below: I agree that with the added OR then this is also correct for termination.